numpy/core/_exceptions.py

```diff
@@ -122,21 +122,105 @@ class TooHardError(RuntimeError):
 
 @set_module('numpy')
 class AxisError(ValueError, IndexError):
-    """ Axis supplied was invalid. """
+    """Axis supplied was invalid.
+
+    Raised whenever an ``axis`` parameter is specified that is larger than
+    the number of array dimensions.
+
+    .. versionadded:: 1.22
+
+        Added the ``axis`` and ``ndim`` attributes.
+
+    Parameters
+    ----------
+    axis : int or str
+        The out of bounds axis or a custom exception message.
+        If an axis is provided, then `ndim` should be specified as well.
+    ndim : int, optional
+        The number of array dimensions.
+    msg_prefix : str, optional
+        A prefix for the exception message.
+
+    Attributes
+    ----------
+    axis : int, optional
+        The out of bounds axis or ``None`` if a custom exception
+        message was provided. This should be the axis as passed by
+        the user, before any normalization to resolve negative indices.
+    ndim : int, optional
+        The number of array dimensions or ``None`` if a custom exception
+        message was provided.
+
+    Examples
+    --------
+    >>> array_1d = np.arange(10)
+    >>> np.cumsum(array_1d, axis=1)
+    Traceback (most recent call last):
+      ...
+    numpy.AxisError: axis 1 is out of bounds for array of dimension 1
+
+    Negative axes are preserved:
+
+    >>> np.cumsum(array_1d, axis=-2)
+    Traceback (most recent call last):
+      ...
+    numpy.AxisError: axis -2 is out of bounds for array of dimension 1
+
+    The class constructor generally takes the axis and arrays'
+    dimensionality as arguments:
+
+    >>> np.AxisError(2, 1, msg_prefix='error')
+    AxisError('error: axis 2 is out of bounds for array of dimension 1')
+
+    Alternatively, a custom exception message can be passed:
+
+    >>> np.AxisError('Custom error message')
+    AxisError('Custom error message')
+
+    """
+
+    __slots__ = ("axis", "ndim")
+
     def __init__(self, axis, ndim=None, msg_prefix=None):
         # single-argument form just delegates to base class
         if ndim is None and msg_prefix is None:
             msg = axis
+            self.axis = None
+            self.ndim = None
 
         # do the string formatting here, to save work in the C code
         else:
             msg = ("axis {} is out of bounds for array of dimension {}"
                    .format(axis, ndim))
             if msg_prefix is not None:
                 msg = "{}: {}".format(msg_prefix, msg)
+            self.axis = axis
+            self.ndim = ndim
 
         super().__init__(msg)
 
+    @classmethod
+    def _reconstruct(cls, axis, ndim, args):
+        """Auxiliary instance constructor used by `__reduce__`.
+
+        Directly assign all instance attributes without further sanitization.
+
+        This method avoids the `__init__` constructor in order to:
```

Comment:
So having looked at it a bit - the reason this is needed is because `Exception.__reduce__` overloads the default `object.__reduce__`. The normal `object.__reduce__` does not call `__init__` anyway.